### ISSUE TITLE
[Style/FileName] Should only require snake case for .rb/.rake/etc. files, not Gemfile, Rakefile, etc.

### ISSUE THREAD

**Issue by Dorian** [rubocop-hq/rubocop#4300]
Requires `Gemfile` to be written `gemfile` :) Related to #3997

--------

## Expected behavior

- No errors

## Actual behavior

```ruby
Gemfile:1:1: C: The name of this source file (Gemfile) should use snake_case.
source 'https://rubygems.org'
^
```

## Steps to reproduce the problem

Have a `Gemfile` file.

## RuboCop version

```
$ rubocop -V
0.48.1 (using Parser 2.4.0.0, running on ruby 2.4.1 x86_64-darwin16)
```

**Comment by Dorian** [comment]
Actually fixed on master #4250

**Comment by Dorian** [comment]
Actually got the same error when using the latest version from master (e.g. `github: 'bbatsov/rubocop'`)

**Comment by tejasbubane** [comment]
There is a spec [here](https://github.com/bbatsov/rubocop/blob/master/spec/rubocop/cop/style/file_name_spec.rb#L107-L117) which requires `Gemfile` to be specified in `AllCops/Include` which is done by default [here](https://github.com/bbatsov/rubocop/blob/dcb3f160a2e23218a7826aadd0e64baefd6d7521/config/default.yml#L45). Is your config causing the issue?

**Comment by perlun** [comment]
I ran into this very issue today as well. Here is my `.rubocop.yml`:

```yaml
AllCops:
  Exclude:
    - bin/*
  Include:
    - bin/mybinary
```

Doing this starts triggering these errors:

```
$ bundle exec rubocop
Inspecting 1 file
C

Offenses:

Gemfile:1:1: C: Naming/FileName: The name of this source file (Gemfile) should use snake_case.
# frozen_string_literal: true
^

1 file inspected, 1 offense detected
```

If I add `Gemfile` or `**/Gemfile` to the `Include` list, it works correctly.

The spec link @tejasbubane referred to is broken, but [here](https://github.com/bbatsov/rubocop/blob/dcb3f160a2e23218a7826aadd0e64baefd6d7521/spec/rubocop/cop/style/file_name_spec.rb#L107-L117) is the fixed link (replacing `master` in the URL with the exact commit he was speaking about.) Do we have a negative spec testing that _not_ including this file in the `Include` list breaks things?

#4250 explains _why_ it behaves like this:

```
Style/FileName:
  # File names listed in `AllCops:Include` are excluded by default. Add extra
  # excludes here.
```

I find the current behavior a bit odd, but maybe it's just me. 😉

**Comment by Drenmi** [comment]
```yaml
AllCops:
  Include:
    - bin/mybinary
```

This overwrites the default `AllCops/Include`. To answer the follow up question: Yes. It has to work like this, or you wouldn't be able to reconfigure cops, removing defaults. 🙂

**Comment by perlun** [comment]
> Yes. It has to work like this, or you wouldn't be able to reconfigure cops, removing defaults. 🙂

That part is fine. But should really `Style/FileName` be **dependent** on a particular `AllCops:Include` setting? That's the part I find debatable.

**Comment by Drenmi** [comment]
> But should really Style/FileName be dependent on a particular AllCops:Include setting? That's the part I find debatable.

That's a good question. I personally think this is bordering on "too clever", and breaking expectations as a result. But that is at least partly subjective. On the other hand we haven't had anyone else open an issue, as I can recall. Think it's a judgement call. Thoughts @pocke, @koic, @rrosenblum, @jonas054?

**Comment by bbatsov** [comment]
I'm closing this ticket due to no recent activity. Feel free to re-open it if you ever come back to it.
